File: src/selenium/browser/process.py
"""یک process مستقل Chrome را برای یک profile مدیریت می‌کند."""

import asyncio
import json
import os
import shutil
from pathlib import Path
from typing import BinaryIO, Protocol
from shared import upstream_value


class ChromeSpec(Protocol):
    browser_id: str


class ChromeProcess:
    def __init__(self, spec: ChromeSpec, port: int, index: int):
        self.spec = spec
        self.port = port
        self.index = index
        self.process: asyncio.subprocess.Process | None = None
        self._log: BinaryIO | None = None

    @property
    def cdp_url(self) -> str:
        return f"http://127.0.0.1:{self.port}"

    async def start(self) -> None:
        profile = self._browser_path("profiles")
        extension = self._browser_path("extensions")
        self._reset_directory(profile)
        self._reset_directory(extension)
        shutil.copytree(self._extension_source(), extension, dirs_exist_ok=True)
        self._write_extension_config(extension)

        runtime = Path(os.getenv("SELENIUM_RUNTIME_DIR", "/app/selenium/runtime"))
        runtime.mkdir(parents=True, exist_ok=True)
        self._log = (runtime / f"chrome-{self.spec.browser_id}.log").open("ab")
        self.process = await asyncio.create_subprocess_exec(
            os.getenv("CHROME_EXECUTABLE", "/usr/bin/google-chrome"),
            *self._arguments(profile, extension),
            stdout=self._log,
            stderr=asyncio.subprocess.STDOUT,
        )
        await self._wait_until_ready()

    async def stop(self) -> None:
        if self.process and self.process.returncode is None:
            self.process.terminate()
            try:
                await asyncio.wait_for(self.process.wait(), timeout=5)
            except TimeoutError:
                self.process.kill()
                await self.process.wait()
        if self._log:
            self._log.close()
        self.process = None
        self._log = None

    def _arguments(self, profile: Path, extension: Path) -> list[str]:
        arguments = [
            "--no-sandbox",
            "--disable-gpu",
            "--no-first-run",
            "--no-default-browser-check",
            "--disable-search-engine-choice-screen",
            "--remote-debugging-address=127.0.0.1",
            f"--remote-debugging-port={self.port}",
            f"--user-data-dir={profile}",
            f"--disable-extensions-except={extension}",
            f"--load-extension={extension}",
            f"--window-position={20 + self.index * 30},{20 + self.index * 30}",
        ]
        proxy = os.getenv("LAB_PROXY_URL", "").strip()
        if proxy:
            arguments.extend([
                f"--proxy-server={proxy}",
                "--proxy-bypass-list=127.0.0.1;localhost;<-loopback>",
            ])
        return [*arguments, "about:blank"]

    async def _wait_until_ready(self) -> None:
        for _attempt in range(100):
            if self.process and self.process.returncode is not None:
                raise RuntimeError(
                    f"Chrome {self.spec.browser_id} exited with {self.process.returncode}"
                )
            try:
                _reader, writer = await asyncio.open_connection("127.0.0.1", self.port)
                writer.close()
                await writer.wait_closed()
                return
            except OSError:
                await asyncio.sleep(0.1)
        raise RuntimeError(f"Chrome {self.spec.browser_id} CDP endpoint did not become ready")

    def _browser_path(self, category: str) -> Path:
        root = Path(os.getenv("CHROME_RUNTIME_DIR", "/tmp/aistudio-browsers"))
        return root / category / self.spec.browser_id

    @staticmethod
    def _reset_directory(path: Path) -> None:
        shutil.rmtree(path, ignore_errors=True)
        path.mkdir(parents=True, exist_ok=True)

    @staticmethod
    def _extension_source() -> Path:
        return Path(os.getenv("EXTENSION_SOURCE_DIR", "/app/extension"))

    def _write_extension_config(self, extension: Path) -> None:
        config = {
            "browserId": self.spec.browser_id,
            "factoryOrigin": os.getenv("FACTORY_ORIGIN", "http://127.0.0.1:3345"),
            "pageMatch": f'{upstream_value("aistudio", "origin")}/*',
        }
        target = extension / "config" / "runtime-config.js"
        target.parent.mkdir(parents=True, exist_ok=True)
        script = f"globalThis.AISTUDIO_BRIDGE_CONFIG = {json.dumps(config)};\n"
        target.write_text(script, encoding="utf-8")

File: internal/chromeprocess/process.go
package chromeprocess

import (
	"encoding/json"
	"fmt"
	"net"
	"os"
	"os/exec"
	"path/filepath"
	"time"

	"github.com/hamed/aistudio-api/internal/aistudio"
)

type Process struct {
	BrowserID   string
	Port, Index int
	command     *exec.Cmd
}

func NewProcess(browserID string, port, index int) *Process {
	return &Process{BrowserID: browserID, Port: port, Index: index}
}
func (p *Process) CDPURL() string { return fmt.Sprintf("http://127.0.0.1:%d", p.Port) }
func (p *Process) Running() bool  { return p.command != nil && p.command.Process != nil }
func (p *Process) Start() error {
	if p.Running() {
		return nil
	}
	runtime := defaultValue(os.Getenv("CHROME_RUNTIME_DIR"), "/app/browser-profiles")
	profile := filepath.Join(runtime, "profiles", p.BrowserID)
	source := defaultValue(os.Getenv("EXTENSION_SOURCE_DIR"), "/app/extension")
	extension, err := extensionDirectory(source, runtime, p.BrowserID)
	if err != nil {
		return err
	}
	if err := os.MkdirAll(profile, 0755); err != nil {
		return err
	}
	for _, name := range []string{"SingletonLock", "SingletonCookie", "SingletonSocket"} {
		_ = os.Remove(filepath.Join(profile, name))
	}
	if err := copyExtension(source, extension, p.BrowserID); err != nil {
		return err
	}
	args := []string{"--no-sandbox", "--disable-gpu", "--no-first-run", "--no-default-browser-check", "--disable-search-engine-choice-screen", "--remote-debugging-address=127.0.0.1", fmt.Sprintf("--remote-debugging-port=%d", p.Port), "--user-data-dir=" + profile, "--disable-extensions-except=" + extension, "--load-extension=" + extension, "about:blank"}
	if proxy := os.Getenv("LAB_PROXY_URL"); proxy != "" {
		args = append(args, "--proxy-server="+proxy, "--proxy-bypass-list=127.0.0.1;localhost;<-loopback>")
	}
	p.command = exec.Command(defaultValue(os.Getenv("CHROME_EXECUTABLE"), "/usr/bin/google-chrome"), args...)
	if err := p.command.Start(); err != nil {
		return err
	}
	return p.waitReady()
}

func extensionDirectory(source, runtime, browserID string) (string, error) {
	data, err := os.ReadFile(filepath.Join(source, "manifest.json"))
	if err != nil {
		return "", err
	}
	var manifest struct {
		Version string `json:"version"`
	}
	if err := json.Unmarshal(data, &manifest); err != nil {
		return "", fmt.Errorf("read extension manifest: %w", err)
	}
	if manifest.Version == "" {
		return "", fmt.Errorf("extension manifest has no version")
	}
	return filepath.Join(runtime, "extensions", browserID, manifest.Version), nil
}
func (p *Process) Stop() {
	if !p.Running() {
		return
	}
	_ = p.command.Process.Kill()
	_, _ = p.command.Process.Wait()
	p.command = nil
}
func (p *Process) waitReady() error {
	for attempt := 0; attempt < 100; attempt++ {
		connection, err := net.DialTimeout("tcp", fmt.Sprintf("127.0.0.1:%d", p.Port), 200*time.Millisecond)
		if err == nil {
			_ = connection.Close()
			return nil
		}
		time.Sleep(100 * time.Millisecond)
	}
	return fmt.Errorf("Chrome %s CDP endpoint did not become ready", p.BrowserID)
}
func copyExtension(source, target, browserID string) error {
	_ = os.RemoveAll(target)
	if err := os.MkdirAll(target, 0755); err != nil {
		return err
	}
	err := filepath.Walk(source, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		relative, _ := filepath.Rel(source, path)
		destination := filepath.Join(target, relative)
		if info.IsDir() {
			return os.MkdirAll(destination, 0755)
		}
		data, err := os.ReadFile(path)
		if err != nil {
			return err
		}
		if err := os.MkdirAll(filepath.Dir(destination), 0755); err != nil {
			return err
		}
		return os.WriteFile(destination, data, 0644)
	})
	if err != nil {
		return err
	}
	upstream, err := aistudio.LoadUpstream()
	if err != nil {
		return err
	}
	config, _ := json.Marshal(map[string]string{"browserId": browserID, "factoryOrigin": defaultValue(os.Getenv("FACTORY_ORIGIN"), "http://127.0.0.1:3345"), "pageMatch": upstream.AIStudio["origin"] + "/*", "chatgptPageMatch": "https://chatgpt.com/*"})
	return os.WriteFile(filepath.Join(target, "config", "runtime-config.js"), []byte("globalThis.AISTUDIO_BRIDGE_CONFIG = "+string(config)+";\n"), 0644)
}
func defaultValue(value, fallback string) string {
	if value == "" {
		return fallback
	}
	return value
}

var _ = aistudio.RuntimeConfig{}
